File: src/backends/dx11/Texture.cpp
#include "pch.h"
#include "Texture.h"
#include "Device.h"

texture_array * rd_create_texture_array(device * dev, const texture_array_params * params)
{
    HRESULT hr;
    D3D11_SUBRESOURCE_DATA *initial_data = nullptr;
    std::vector<D3D11_SUBRESOURCE_DATA> subresources;
    std::unique_ptr<texture_array> tary{ new texture_array };

    tary->streaming = params->streaming;
    tary->pixel_art = params->pixel_art;
    tary->width = params->sprite_width;
    tary->height = params->sprite_height;

    tary->textures.reserve(params->sprite_count);
    for (uint32_t i = 0; i < params->sprite_count; ++i)
    {
        texture tex;
        tex.array = tary.get();
        tex.index = i;
        tary->textures.push_back(tex);
    }

    if (params->buffers)
    {
        subresources.reserve(params->sprite_count);
        for (uint32_t i = 0; i < params->sprite_count; ++i)
        {
            D3D11_SUBRESOURCE_DATA subres;
            subres.pSysMem = params->buffers[i];
            subres.SysMemPitch = params->sprite_width * 4;
            subres.SysMemSlicePitch = 0;
            subresources.push_back(subres);
        }
        initial_data = subresources.data();
    }

    D3D11_TEXTURE2D_DESC desc;
    desc.ArraySize = params->sprite_count;
    desc.BindFlags = D3D11_BIND_SHADER_RESOURCE;
    desc.CPUAccessFlags = params->streaming ? D3D11_CPU_ACCESS_WRITE : 0;
    desc.Format = DXGI_FORMAT_R8G8B8A8_UNORM;
    desc.Width = params->sprite_width;
    desc.Height = params->sprite_height;
    desc.MiscFlags = 0;
    desc.MipLevels = 1;
    desc.SampleDesc = { 1, 0 };
    desc.Usage = params->streaming ? D3D11_USAGE_DYNAMIC : D3D11_USAGE_DEFAULT;
    hr = dev->d3d_device->CreateTexture2D(&desc, initial_data, &tary->buffer);
    if (FAILED(hr))
        return append_error_and_ret(set_error_and_ret(hr), "Failed to create Texture2D");

    D3D11_SHADER_RESOURCE_VIEW_DESC srvdesc;
    srvdesc.Format = desc.Format;
    srvdesc.ViewDimension = D3D11_SRV_DIMENSION_TEXTURE2DARRAY;
    srvdesc.Texture2DArray.ArraySize = desc.ArraySize;
    srvdesc.Texture2DArray.FirstArraySlice = 0;
    srvdesc.Texture2DArray.MipLevels = 1;
    srvdesc.Texture2DArray.MostDetailedMip = 0;
    hr = dev->d3d_device->CreateShaderResourceView(tary->buffer, &srvdesc, &tary->srv);
    if (FAILED(hr))
        return append_error_and_ret(set_error_and_ret(hr), "Failed to create ShaderResourceView");

    return tary.release();
}

void rd_free_texture_array(texture_array * set)
{
    delete set;
}

void rd_get_texture_array_size(const texture_array * set, uint32_t * width, uint32_t * height)
{
    *width = set->width;
    *height = set->height;
}

uint32_t rd_get_texture_array_count(const texture_array * set)
{
    return (uint32_t)set->textures.size();
}

bool rd_is_texture_array_streaming(const texture_array * set)
{
    return set->streaming;
}

bool rd_is_texture_array_pixel_art(const texture_array * set)
{
    return set->pixel_art;
}

void rd_set_texture_array_pixel_art(texture_array * set, bool pa)
{
    set->pixel_art = pa;
}

texture * rd_get_texture(texture_array * set, uint32_t index)
{
    if (index >= set->textures.size())
        return set_error_and_ret("Index out of bounds");

    return &set->textures[index];
}

texture_array * rd_get_texture_array(texture * texture)
{
    return texture->array;
}

bool rd_update_texture(device *dev, texture *texture, const uint8_t *data, size_t len)
{
    auto ary = texture->array;
    if (len != ary->width * ary->height * 4)
        return set_error_and_ret(false, "Wrong buffer size for updating a texture");

    if (ary->streaming)
    {
        HRESULT hr;
        D3D11_MAPPED_SUBRESOURCE subres;
        hr = dev->d3d_context->Map(ary->buffer, texture->index, D3D11_MAP_WRITE_DISCARD, 0, &subres);
        if (FAILED(hr))
            return set_error_and_ret(false, "Failed to map texture buffer");

        for (uint32_t y = 0; y < ary->height; ++y)
        {
            const uint8_t *in = &data[y * ary->width * 4];
            uint8_t *out = ((uint8_t *)subres.pData) + y * subres.RowPitch;

            memcpy(out, in, ary->width * 4);
        }

        dev->d3d_context->Unmap(ary->buffer, texture->index);
    }
    else
    {
        D3D11_BOX box;
        box.front = 0;
        box.left = 0;
        box.top = 0;
        box.back = 1;
        box.right = ary->width;
        box.bottom = ary->height;
        dev->d3d_context->UpdateSubresource(
            ary->buffer,
            texture->index,
            &box,
            data,
            ary->width * 4,
            0
        );
    }

    return true;
}

File: src/engine/graphics/texture.lua
local ffi = require("engine.graphics.renderer")
local rd_err = require("engine.graphics.error")

local C = ffi.C
local check_ptr = rd_err.check_ptr
local ffi_new = ffi.new
local ffi_cast = ffi.cast
local ffi_gc = ffi.gc
local ffi_string = ffi.string

local tparams = ffi.typeof("struct texture_array_params")

local TextureArray_t = ffi.typeof("struct{texture_array *tary;}")
local TextureArray = {}
local TextureArray_mt = { __index = TextureArray }
local TextureArray_ct

local Texture_t = ffi.typeof("struct{texture *tex;}")
local Texture = {}
local Texture_mt = { __index = Texture }
local Texture_ct

function TextureArray.build_placeholder(colors, device)
    device = device or colors.device
    local height = #colors
    local width = #colors[1]
    local buffer = ffi_cast("uint8_t(*)[4]", C.malloc(width * height * 4))
    local buffers = ffi_new("const uint8_t *[1]", buffer[0])

    for i, row in ipairs(colors) do
        for j, color in ipairs(row) do
            local idx = (i - 1) * width + j - 1
            buffer[idx][0] = color.r * 255
            buffer[idx][1] = color.g * 255
            buffer[idx][2] = color.b * 255
            buffer[idx][3] = color.a * 255
        end
    end
    
    local params = ffi_new(tparams)
    params.streaming = false
    params.sprite_count = 1
    params.sprite_width = width
    params.sprite_height = height
    params.buffers = buffers
    params.pixel_art = true

    return TextureArray_ct(device, params)
end

function TextureArray_mt.__new(tp, dev, params)
    local tary = check_ptr(__rd.rd_create_texture_array(dev.dev, params))
    return ffi_new(tp, tary)
end

function TextureArray_mt:__gc()
    self:destroy()
end

function TextureArray:destroy()
    if self.tary ~= nil then
        __rd.rd_free_texture_array(self.tary)
        self.tary = nil
    end
end

function TextureArray:get(i)
    local tex = check_ptr(__rd.rd_get_texture(self.tary, i))
    return Texture_ct(tex)
end

TextureArray_ct = ffi.metatype(TextureArray_t, TextureArray_mt)
Texture_ct = ffi.metatype(Texture_t, Texture_mt)

return {
    TextureArray = TextureArray,
}
